rust/pattern/variable/thing_variable.rs

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright (C) 2022 Vaticle
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+use crate::pattern::*;
+use crate::write_joined;
+use std::fmt;
+use std::fmt::{Display, Write};
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct ThingVariable {
+    pub reference: Reference,
+    pub isa: Option<IsaConstraint>,
+    pub has: Vec<HasConstraint>,
+    pub value: Option<ValueConstraint>,
+    pub relation: Option<RelationConstraint>,
+}
```

Comment:
There isn't a type system level distinction between a Relation, Attribute, or a generic `ThingVariable`, and I'm not sure there should be.